ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFDateDiff.java

```diff
@@ -64,121 +66,46 @@
-    if (arguments.length != 2) {
-      throw new UDFArgumentLengthException(
-        "datediff() requires 2 argument, got " + arguments.length);
-    }
-    inputConverter1 = checkArguments(arguments, 0);
-    inputConverter2 = checkArguments(arguments, 1);
-    inputType1 = ((PrimitiveObjectInspector) arguments[0]).getPrimitiveCategory();
-    inputType2 = ((PrimitiveObjectInspector) arguments[1]).getPrimitiveCategory();
-    ObjectInspector outputOI = PrimitiveObjectInspectorFactory.writableIntObjectInspector;
-    return outputOI;
+    checkArgsSize(arguments,2,2);
+    checkArgPrimitive(arguments, 0);
+    checkArgPrimitive(arguments, 1);
+    checkArgGroups(arguments, 0, tsInputTypes, STRING_GROUP, DATE_GROUP);
+    checkArgGroups(arguments, 1, tsInputTypes, STRING_GROUP, DATE_GROUP);
+    obtainTimestampConverter(arguments,0, tsInputTypes, tsConverters);
+    obtainTimestampConverter(arguments,1, tsInputTypes, tsConverters);
+    return PrimitiveObjectInspectorFactory.writableIntObjectInspector;
   }
 
   @Override
   public IntWritable evaluate(DeferredObject[] arguments) throws HiveException {
-    output = evaluate(convertToDate(inputType1, inputConverter1, arguments[0]),
-      convertToDate(inputType2, inputConverter2, arguments[1]));
-    return output;
-  }
 
-  @Override
-  public String getDisplayString(String[] children) {
-    return getStandardDisplayString("datediff", children);
-  }
+    Timestamp ts1 = getTimestampValue(arguments, 0, tsConverters);
+    Timestamp ts2 = getTimestampValue(arguments, 1, tsConverters);
 
-  @Nullable
-  private Date convertToDate(PrimitiveCategory inputType, Converter converter, DeferredObject argument)
-    throws HiveException {
-    assert(converter != null);
-    assert(argument != null);
-    if (argument.get() == null) {
-      return null;
-    }
-    switch (inputType) {
-    case STRING:
-    case VARCHAR:
-    case CHAR: {
-      String dateString = converter.convert(argument.get()).toString();
-      Date date = DateParser.parseDate(dateString);
-      if (date != null) {
-        return date;
-      }
-      Timestamp ts = PrimitiveObjectInspectorUtils.getTimestampFromString(dateString);
-      if (ts != null) {
-        return Date.ofEpochMilli(ts.toEpochMilli());
-      }
+    if (ts1 == null || ts2 == null) {
       return null;
     }
-    case TIMESTAMP:
-      Timestamp ts = ((TimestampWritableV2) converter.convert(argument.get()))
-        .getTimestamp();
-      return Date.ofEpochMilli(ts.toEpochMilli());
-    case DATE:
-      DateWritableV2 dw = (DateWritableV2) converter.convert(argument.get());
-      return dw.get();
-    case TIMESTAMPLOCALTZ:
-      TimestampTZ tsz = ((TimestampLocalTZWritable) converter.convert(argument.get()))
-          .getTimestampTZ();
-      return Date.ofEpochMilli(tsz.getEpochSecond() * 1000l);
-    default:
-      throw new UDFArgumentException(
-        "TO_DATE() only takes STRING/TIMESTAMP/TIMESTAMPLOCALTZ types, got " + inputType);
```

Comment:
I didn't see any negative tests verifying that an error is raised when types are illegal. Would it be possible to add some tests to make sure that the refactoring doesn't break something?

Reply:
Yes we have couple of qtest which checks both positive and negative. Include command like "DESCRIBE FUNCTION EXTENDED datediff";